[Bug] deviceshifu-http-mqtt:v0.0.6 crashes

**Describe the bug**

Use ***Shifu*** to create a MQTT ***deviceShifu*** will fail using `v0.0.6`.

**To Reproduce**

```
git clone https://github.com/Edgenesis/shifu.git
cd shifu/examples/mqttDeviceShifu/mqtt_deploy
```

Modify `spec.address` and `protocolSettings.MQTTSetting.MQTTTopic` to my MQTT broker's IP address and topic.

```
$ sudo kubectl apply -f .

$ sudo kubectl get pods -A
NAMESPACE            NAME                                            READY   STATUS    RESTARTS     AGE
deviceshifu          deviceshifu-mqtt-deployment-74bfc6d877-4bxhn    0/1     Error     1 (9s ago)   14s

$ sudo kubectl describe pod deviceshifu-mqtt-deployment-74bfc6d877-4bxhn -n deviceshifu
Name:             deviceshifu-mqtt-deployment-74bfc6d877-4bxhn
Namespace:        deviceshifu
Priority:         0
Service Account:  edgedevice-sa
Node:             kind-control-plane/172.18.0.2
Start Time:       Thu, 01 Sep 2022 14:21:07 +0800
Labels:           app=deviceshifu-mqtt-deployment
                  pod-template-hash=74bfc6d877
Annotations:      <none>
Status:           Running
IP:               10.244.0.8
IPs:
  IP:           10.244.0.8
Controlled By:  ReplicaSet/deviceshifu-mqtt-deployment-74bfc6d877
Containers:
  deviceshifu-http:
    Container ID:   containerd://c595e4a2e1eae64ae0c097f679d30a713b2df07039f16c64bfe40d658f7aaa4d
    Image:          edgehub/deviceshifu-http-mqtt:v0.0.6
    Image ID:       docker.io/edgehub/deviceshifu-http-mqtt@sha256:329ad1beaa74d6fa8c586ec8055e1e8db584a0595039e50d9917f496b3f0bfbd
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Thu, 01 Sep 2022 14:22:05 +0800
      Finished:     Thu, 01 Sep 2022 14:22:10 +0800
    Ready:          False
    Restart Count:  3
    Environment:
      EDGEDEVICE_NAME:       edgedevice-mqtt
      EDGEDEVICE_NAMESPACE:  devices
    Mounts:
      /etc/edgedevice/config from deviceshifu-config (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-x27b9 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  deviceshifu-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      mqtt-configmap-0.0.1
    Optional:  false
  kube-api-access-x27b9:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  87s                default-scheduler  Successfully assigned deviceshifu/deviceshifu-mqtt-deployment-74bfc6d877-4bxhn to kind-control-plane
  Normal   Pulled     29s (x4 over 87s)  kubelet            Container image "edgehub/deviceshifu-http-mqtt:v0.0.6" already present on machine
  Normal   Created    29s (x4 over 87s)  kubelet            Created container deviceshifu-http
  Normal   Started    29s (x4 over 87s)  kubelet            Started container deviceshifu-http
  Warning  BackOff    11s (x5 over 75s)  kubelet            Back-off restarting failed container

$ sudo kubectl logs deviceshifu-mqtt-deployment-74bfc6d877-4bxhn -n deviceshifu
Password:
2022/09/01 06:27:26 MQTT Server Address is empty, use address instead
2022/09/01 06:27:26 Connected
2022/09/01 06:27:26 Subscribed to topic: topic0
2022/09/01 06:27:26 updating device edgedevice-mqtt status to: Pending
2022/09/01 06:27:26 deviceShifu edgedevice-mqtt started
2022/09/01 06:27:26 Wait 5 seconds before updating status
deviceShifu edgedevice-mqtt's http server started
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xfa4e4f]

goroutine 11 [running]:
github.com/edgenesis/shifu/deviceshifu/pkg/deviceshifubase.getTelemetryCollectionServiceMap(0xc000098cf0)
	/shifu/deviceshifu/pkg/deviceshifubase/telemetry_service.go:49 +0xaf
github.com/edgenesis/shifu/deviceshifu/pkg/deviceshifubase.(*DeviceShifuBase).StartTelemetryCollection(0xc000098cf0, 0x0?)
	/shifu/deviceshifu/pkg/deviceshifubase/deviceshifubase.go:222 +0x6f
created by github.com/edgenesis/shifu/deviceshifu/pkg/deviceshifubase.(*DeviceShifuBase).Start
	/shifu/deviceshifu/pkg/deviceshifubase/deviceshifubase.go:258 +0x119
```


**Expected behavior**

I use `deviceshifu-http-mqtt:v0.0.1` and it works!

```
$ sudo kubectl get pods -A -w
deviceshifu          deviceshifu-mqtt-deployment-65fd55674b-ptcq7    1/1     Running       0          7s
$ sudo kubectl exec -it nginx -- bash
root@nginx:/# curl deviceshifu-mqtt.deviceshifu.svc.cluster.local/mqtt_data
{"mqtt_message":"哈哈哈","mqtt_receive_timestamp":"2022-09-01 06:55:15.60403948 +0000 UTC m=+47.354876899"}
```

**Server (please complete the following information):**

 - Shifu Version: v0.0.6
 - kind 0.14.0 on macOS 12.5.1
 - CPU Architecture x86-64

--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ vet: ## Run go vet against code.
 
 .PHONY: test
 test: fmt envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -race -coverprofile=coverage.out -covermode=atomic ./...
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -v -race -coverprofile=coverage.out -covermode=atomic ./...
 
 buildx-push-image-deviceshifu-http-http:
 	docker buildx build --platform=linux/amd64,linux/arm64,linux/arm -f ${PROJECT_ROOT}/dockerfiles/Dockerfile.deviceshifuHTTP \

--- a/examples/mqttDeviceShifu/steps.md
+++ b/examples/mqttDeviceShifu/steps.md
@@ -2,7 +2,7 @@
 ### Option 1
 To start MQTT broker in local machine:
 
-```
+```bash
 # install mosquitto server and client
 sudo apt-add-repository ppa:mosquitto-dev/mosquitto-ppa
 sudo apt-get install mosquitto mosquitto-clients
@@ -18,7 +18,7 @@ log_dest stdout
 
 restart mosquitto service:
 
-```
+```bash
 sudo service mosquitto restart
 ```
 
@@ -26,7 +26,7 @@ sudo service mosquitto restart
 ### Option 2
 Deploy the MQTT broker in Kubernetes:
 
-```
+```bash
 kubectl apply -f /shifu/examples/mqttDeviceShifu/mqtt_broker
 # Forward Kubernetes svc to localhost:
 kubectl port-forward svc/mosquitto-service -n devices 1883:1883 --address='0.0.0.0'
@@ -35,31 +35,34 @@ kubectl port-forward svc/mosquitto-service -n devices 1883:1883 --address='0.0.0
 ### Testing:
 Deploy MQTT deviceshifu YAMLs:
 
-```
+```bash
 kubectl apply -f examples/mqttDeviceShifu/mqtt_deploy/
 ```
 
 Publish some ramdom data to the topic:
-```
+
+```bash
 mosquitto_pub -h 172.28.15.229 -d -p 1883 -t /test/test -m "test2333"
 ```
 
 Test using an nginx pod
-```
-kubectl run nginx --image=nginx
-kubectl exec -it nginx -- bash
-curl edgedevice-mqtt/mqtt_data
+
+```bash
+kubectl run nginx --image=nginx -n deviceshifu
+kubectl -n deviceshifu exec -it nginx -- bash
+curl deviceshifu-mqtt/mqtt_data
 ```
 
 output:
-```
-root@nginx:/# curl edgedevice-mqtt/mqtt_data
+
+```bash
+root@nginx:/# curl deviceshifu-mqtt/mqtt_data
 {"mqtt_message":"test2333","mqtt_receive_timestamp":"2022-04-29 08:57:49.9492744 +0000 UTC m=+75.407609501"}
 ```
 
 For debugging, subscribe to your machine's IP (change accordingly):
 
-```
+```bash
 mosquitto_sub -h 172.28.15.229 -p 1883 -t /test/test -d
 ```
 

--- a/pkg/deviceshifu/deviceshifuHTTP/deviceshifuHTTP_test.go
+++ b/pkg/deviceshifu/deviceshifuHTTP/deviceshifuHTTP_test.go
@@ -1,7 +1,6 @@
 package deviceshifuHTTP
 
 import (
-	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
 	"io"
 	"log"
 	"net/http"
@@ -10,6 +9,9 @@ import (
 	"sort"
 	"strings"
 	"testing"
+
+	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
+	"github.com/edgenesis/shifu/pkg/deviceshifu/utils"
 
 	"k8s.io/apimachinery/pkg/util/wait"
 )
@@ -68,7 +70,7 @@ func TestDeviceHealthHandler(t *testing.T) {
 		t.Errorf("DeviceShifu.Start failed due to: %v", err.Error())
 	}
 
-	resp, err := http.Get("http://localhost:8080/health")
+	resp, err := utils.RetryAndGetHTTP("http://localhost:8080/health", 3)
 	if err != nil {
 		t.Errorf("HTTP GET returns an error %v", err.Error())
 	}

--- a/pkg/deviceshifu/deviceshifuMQTT/deviceshifuMQTT.go
+++ b/pkg/deviceshifu/deviceshifuMQTT/deviceshifuMQTT.go
@@ -3,12 +3,13 @@ package deviceshifuMQTT
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
-	"github.com/edgenesis/shifu/pkg/k8s/api/v1alpha1"
 	"log"
 	"net/http"
 	"strings"
 	"time"
+
+	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
+	"github.com/edgenesis/shifu/pkg/k8s/api/v1alpha1"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
 )
@@ -224,8 +225,9 @@ func (ds *DeviceShifu) collectMQTTTelemetry() (bool, error) {
 				return false, fmt.Errorf("Device %v does not have an address", ds.base.Name)
 			}
 
-			if telemetrySettings.DeviceShifuTelemetryUpdateIntervalInMilliseconds == nil {
-				*telemetrySettings.DeviceShifuTelemetryUpdateIntervalInMilliseconds = DEFAULT_UPDATE_INTERVAL_MS
+			if interval := telemetrySettings.DeviceShifuTelemetryUpdateIntervalInMilliseconds; interval == nil {
+				var telemetryUpdateIntervalInMilliseconds = DEFAULT_UPDATE_INTERVAL_MS
+				telemetrySettings.DeviceShifuTelemetryUpdateIntervalInMilliseconds = &telemetryUpdateIntervalInMilliseconds
 			}
 
 			nowTime := time.Now()

--- a/pkg/deviceshifu/deviceshifuMQTT/deviceshifuMQTT_test.go
+++ b/pkg/deviceshifu/deviceshifuMQTT/deviceshifuMQTT_test.go
@@ -1,7 +1,6 @@
 package deviceshifuMQTT
 
 import (
-	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
 	"io"
 	"log"
 	"net/http"
@@ -10,6 +9,9 @@ import (
 	"sort"
 	"strings"
 	"testing"
+
+	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
+	"github.com/edgenesis/shifu/pkg/deviceshifu/utils"
 
 	"k8s.io/apimachinery/pkg/util/wait"
 )
@@ -53,7 +55,7 @@ func TestDeviceHealthHandler(t *testing.T) {
 		t.Errorf("DeviceShifu.Start failed due to: %v", err.Error())
 	}
 
-	resp, err := http.Get("http://localhost:8080/health")
+	resp, err := utils.RetryAndGetHTTP("http://localhost:8080/health", 3)
 	if err != nil {
 		t.Errorf("HTTP GET returns an error %v", err.Error())
 	}

--- a/pkg/deviceshifu/deviceshifuOPCUA/deviceshifuOPCUA_test.go
+++ b/pkg/deviceshifu/deviceshifuOPCUA/deviceshifuOPCUA_test.go
@@ -1,12 +1,13 @@
 package deviceshifuOPCUA
 
 import (
-	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
 	"io"
 	"log"
-	"net/http"
 	"os"
 	"testing"
+
+	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
+	"github.com/edgenesis/shifu/pkg/deviceshifu/utils"
 
 	"k8s.io/apimachinery/pkg/util/wait"
 )
@@ -79,7 +80,7 @@ func TestDeviceHealthHandler(t *testing.T) {
 		t.Errorf("DeviceShifu.Start failed due to: %v", err.Error())
 	}
 
-	resp, err := http.Get("http://localhost:8080/health")
+	resp, err := utils.RetryAndGetHTTP("http://localhost:8080/health", 3)
 	if err != nil {
 		t.Errorf("HTTP GET returns an error %v", err.Error())
 	}

--- a/pkg/deviceshifu/deviceshifuSocket/deviceshifuSocket_test.go
+++ b/pkg/deviceshifu/deviceshifuSocket/deviceshifuSocket_test.go
@@ -1,7 +1,6 @@
 package deviceshifuSocket
 
 import (
-	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
 	"io"
 	"log"
 	"net/http"
@@ -10,6 +9,9 @@ import (
 	"sort"
 	"strings"
 	"testing"
+
+	"github.com/edgenesis/shifu/pkg/deviceshifu/deviceshifubase"
+	"github.com/edgenesis/shifu/pkg/deviceshifu/utils"
 
 	"k8s.io/apimachinery/pkg/util/wait"
 )
@@ -51,7 +53,7 @@ func TestDeviceHealthHandler(t *testing.T) {
 		t.Errorf("DeviceShifu.Start failed due to: %v", err.Error())
 	}
 
-	resp, err := http.Get("http://localhost:8080/health")
+	resp, err := utils.RetryAndGetHTTP("http://localhost:8080/health", 3)
 	if err != nil {
 		t.Errorf("HTTP GET returns an error %v", err.Error())
 	}

--- a/pkg/deviceshifu/deviceshifubase/telemetry_service.go
+++ b/pkg/deviceshifu/deviceshifubase/telemetry_service.go
@@ -3,10 +3,11 @@ package deviceshifubase
 import (
 	"context"
 	"fmt"
-	"github.com/edgenesis/shifu/pkg/k8s/api/v1alpha1"
 	"log"
 	"net/http"
 	"time"
+
+	"github.com/edgenesis/shifu/pkg/k8s/api/v1alpha1"
 )
 
 // HTTP header type:
@@ -45,6 +46,10 @@ func getTelemetryCollectionServiceMap(ds *DeviceShifuBase) (map[string]string, e
 	defaultTelemetryCollectionService := ""
 	defaultTelemetryServiceAddress := ""
 	telemetries := ds.DeviceShifuConfig.Telemetries
+	if telemetries.DeviceShifuTelemetrySettings == nil {
+		telemetries.DeviceShifuTelemetrySettings = &DeviceShifuTelemetrySettings{}
+	}
+
 	if telemetries.DeviceShifuTelemetrySettings.DeviceShifuTelemetryDefaultPushToServer != nil {
 		defaultPushToServer = *telemetries.DeviceShifuTelemetrySettings.DeviceShifuTelemetryDefaultPushToServer
 	}

--- a/pkg/deviceshifu/utils/testUtils.go
+++ b/pkg/deviceshifu/utils/testUtils.go
@@ -1,0 +1,30 @@
+package utils
+
+import (
+	"log"
+	"net/http"
+	"time"
+)
+
+func RetryAndGetHTTP(url string, retries int) (*http.Response, error) {
+	var (
+		err      error
+		response *http.Response
+	)
+
+	for retries > 0 {
+		response, err = http.Get(url)
+		if err != nil {
+			log.Println(err)
+			retries -= 1
+			time.Sleep(time.Second * 1)
+			continue
+		}
+
+		if response.StatusCode == http.StatusOK {
+			return response, nil
+		}
+	}
+
+	return nil, err
+}

